File: web/js/anchor/ix/craft/add.ts
import {AnchorProvider, Program, SplToken} from "@project-serum/anchor";
import * as InitPixel from "./init";
import * as Palette from "../../pda/craft/palette-pda";
import * as Pixel from "../../pda/craft/pixel-pda";
import * as PixelIndex from "../../pda/craft/pixel-index-pda";
import * as PixelIndexLookup from "../../pda/craft/pixel-index-lookup-pda";
import {deriveAtaPda} from "../../pda/ata-pda";
import {SPL_ASSOCIATED_TOKEN_PROGRAM_ID, SPL_TOKEN_PROGRAM_ID} from "../../util/constants";
import {SystemProgram, SYSVAR_RENT_PUBKEY} from "@solana/web3.js";
import {getGlobal} from "../../pda/get-global";
import {SRgbStake} from "../../idl/stake";
import {SRgbCraft} from "../../idl/craft";
import {SRgbPaint} from "../../idl/paint";

export interface Input {
    left: Pixel.Seeds,
    right: Pixel.Seeds
}

export async function ix(
    app,
    provider: AnchorProvider,
    programs: {
        stake: Program<SRgbStake>;
        craft: Program<SRgbCraft>;
        paint: Program<SRgbPaint>;
        token: Program<SplToken>
    },
    input: Input
): Promise<void> {
    const leftPixelPda = Pixel.derivePixelPda(
        programs.craft,
        input.left
    );
    const leftPixel = await Pixel.getPixelPda(
        provider,
        programs,
        leftPixelPda
    );
    const rightPixelPda = Pixel.derivePixelPda(
        programs.craft,
        input.right
    );
    const rightPixel = await Pixel.getPixelPda(
        provider,
        programs,
        rightPixelPda
    );
    const dstPixelSeeds = add(
        input.left,
        input.right
    );
    const dstPixelPda = Pixel.derivePixelPda(
        programs.craft,
        dstPixelSeeds
    );
    const dstPixel: Pixel.Pixel = await InitPixel.getOrInit(
        provider,
        programs,
        dstPixelSeeds,
        dstPixelPda
    );
    const dstPalettePda = Palette.derivePalettePda(
        programs.craft,
        {
            authority: provider.wallet.publicKey,
            depth: input.left.depth
        }
    );
    const dstPalette = await Palette.getPalettePda(
        programs.craft,
        dstPalettePda
    );
    const distPixelIndexLookupSeeds: PixelIndexLookup.Seeds = {
        authority: provider.wallet.publicKey,
        r: dstPixelSeeds.r,
        g: dstPixelSeeds.g,
        b: dstPixelSeeds.b,
        depth: dstPixelSeeds.depth
    };
    const dstPixelIndexLookupPda = PixelIndexLookup.derivePixelIndexLookupPda(
        programs.craft,
        distPixelIndexLookupSeeds
    );
    let dstPixelIndexLookup: PixelIndexLookup.PixelIndexLookup;
    try {
        dstPixelIndexLookup = await PixelIndexLookup.getPixelIndexLookupPda(
            programs.craft,
            dstPixelIndexLookupPda
        );
    } catch (error) {
        console.log(error);
        dstPixelIndexLookup = {
            seeds: distPixelIndexLookupSeeds,
            index: dstPalette.indexer + 1
        }
    }
    const dstPixelIndexSeeds = {
        authority: provider.wallet.publicKey,
        depth: input.left.depth,
        index: dstPixelIndexLookup.index
    } as PixelIndex.Seeds;
    const dstPixelIndexPda = PixelIndex.derivePixelIndexPda(
        programs.craft,
        dstPixelIndexSeeds
    )
    const leftPixelMintAta = deriveAtaPda(
        provider.wallet.publicKey,
        leftPixel.mint
    );
    const rightPixelMintAta = deriveAtaPda(
        provider.wallet.publicKey,
        rightPixel.mint
    );
    const dstPixelMintAta = deriveAtaPda(
        provider.wallet.publicKey,
        dstPixel.mint
    );
    await programs
        .craft
        .methods
        .addPixel(
            PixelIndex.toRaw(dstPixelIndexSeeds) as any,
            distPixelIndexLookupSeeds as any
        )
        .accounts(
            {
                leftPixel: leftPixelPda.address,
                rightPixel: rightPixelPda.address,
                dstPixel: dstPixelPda.address,
                dstPixelIndex: dstPixelIndexPda.address,
                dstPixelIndexLookup: dstPixelIndexLookupPda.address,
                dstPalette: dstPalettePda.address,
                leftPixelMint: leftPixel.mint,
                leftPixelMintAta: leftPixelMintAta,
                rightPixelMint: rightPixel.mint,
                rightPixelMintAta: rightPixelMintAta,
                dstPixelMint: dstPixel.mint,
                dstPixelMintAta: dstPixelMintAta,
                payer: provider.wallet.publicKey,
                tokenProgram: SPL_TOKEN_PROGRAM_ID,
                associatedTokenProgram: SPL_ASSOCIATED_TOKEN_PROGRAM_ID,
                systemProgram: SystemProgram.programId,
                rent: SYSVAR_RENT_PUBKEY
            }
        ).rpc();
    await getGlobal(
        app,
        provider,
        programs
    );
}

function add(left: Pixel.Seeds, right: Pixel.Seeds): Pixel.Seeds {
    let max = (2 ** left.depth) - 1;
    let r = Math.min(
        left.r + right.r,
        max
    );
    let g = Math.min(
        left.g + right.g,
        max
    );
    let b = Math.min(
        left.b + right.b,
        max
    );
    return {
        r: r,
        g: g,
        b: b,
        depth: left.depth
    }
}

File: programs/s-rgb-craft/src/ix/craft/math/add.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{burn, Burn, mint_to, MintTo};
use crate::{AddPixel, HasFiveSeeds, Pixel, PixelIndexLookupSeeds, PixelIndexSeeds};
use crate::error::CustomErrors;
use crate::ix::craft::math::{assert_depth, index};

pub fn ix(
    ctx: Context<AddPixel>,
    dst_pixel_index_seeds: PixelIndexSeeds,
    dst_pixel_index_lookup_seeds: PixelIndexLookupSeeds,
) -> Result<()> {
    // get accounts
    let left_pixel = &ctx.accounts.left_pixel;
    let right_pixel = &ctx.accounts.right_pixel;
    let dst_pixel = &ctx.accounts.dst_pixel;
    let dst_pixel_index = &mut ctx.accounts.dst_pixel_index;
    let dst_pixel_index_lookup = &mut ctx.accounts.dst_pixel_index_lookup;
    let dst_palette = &mut ctx.accounts.dst_palette;
    // assert depth
    assert_depth(left_pixel, right_pixel, dst_pixel)?;
    // assert addition
    assert_addition(left_pixel, right_pixel, dst_pixel)?;
    // build signer seeds
    let bump = *ctx.bumps.get(
        "dst_pixel"
    ).unwrap();
    let seed1 = dst_pixel.seed1();
    let seed2 = dst_pixel.seed2();
    let seed3 = dst_pixel.seed3();
    let seed4 = dst_pixel.seed4();
    let seed5 = dst_pixel.seed5();
    let seeds = &[
        seed1.as_slice(),
        seed2.as_slice(),
        seed3.as_slice(),
        seed4.as_slice(),
        seed5.as_slice(),
        &[bump]
    ];
    let signer_seeds = &[&seeds[..]];
    // build burn left-pixel ix
    let burn_left_pixel_cpi_context = CpiContext::new(
        ctx.accounts.token_program.to_account_info(),
        Burn {
            mint: ctx.accounts.left_pixel_mint.to_account_info(),
            from: ctx.accounts.left_pixel_mint_ata.to_account_info(),
            authority: ctx.accounts.payer.to_account_info(),
        },
    );
    // build burn right-pixel ix
    let burn_right_pixel_cpi_context = CpiContext::new(
        ctx.accounts.token_program.to_account_info(),
        Burn {
            mint: ctx.accounts.right_pixel_mint.to_account_info(),
            from: ctx.accounts.right_pixel_mint_ata.to_account_info(),
            authority: ctx.accounts.payer.to_account_info(),
        },
    );
    // build mint dst-pixel ix
    let mint_dst_pixel_cpi_context = CpiContext::new(
        ctx.accounts.token_program.to_account_info(),
        MintTo {
            mint: ctx.accounts.dst_pixel_mint.to_account_info(),
            to: ctx.accounts.dst_pixel_mint_ata.to_account_info(),
            authority: ctx.accounts.dst_pixel.to_account_info(),
        },
    );
    // invoke burn left-pixel ix
    burn(
        burn_left_pixel_cpi_context,
        1,
    )?;
    // invoke burn right-pixel ix
    burn(
        burn_right_pixel_cpi_context,
        1,
    )?;
    // invoke mint dst-pixel ix
    mint_to(
        mint_dst_pixel_cpi_context.with_signer(
            signer_seeds
        ),
        1,
    )?;
    // index
    index(
        dst_pixel.key(),
        dst_pixel_index,
        dst_pixel_index_seeds,
        dst_pixel_index_lookup,
        dst_pixel_index_lookup_seeds,
        dst_palette,
    );
    Ok(())
}

fn assert_addition(left: &Pixel, right: &Pixel, dst: &Pixel) -> Result<()> {
    let max = u32::pow(2, dst.seeds.depth as u32) - 1;
    let r = u32::min(left.seeds.r + right.seeds.r, max);
    let g = u32::min(left.seeds.g + right.seeds.g, max);
    let b = u32::min(left.seeds.b + right.seeds.b, max);
    match (r.eq(&dst.seeds.r), g.eq(&dst.seeds.g), b.eq(&dst.seeds.b)) {
        (true, true, true) => {
            Ok(())
        }
        _ => {
            Err(CustomErrors::InvalidAddition.into())
        }
    }
}
